core/src/main/java/com/linecorp/armeria/internal/DefaultTimeoutController.java

```diff
@@ -78,74 +79,98 @@ public void setTimeoutTask(TimeoutTask timeoutTask) {
      * the {@link #DefaultTimeoutController(TimeoutTask, EventLoop)} before calling this method.
      */
     @Override
-    public void initTimeout(long timeoutMillis) {
-        if (timeoutFuture != null || timeoutTask == null || !timeoutTask.canSchedule()) {
-            // No need to schedule a response timeout if:
-            // - the timeout has been scheduled already,
-            // - the timeout task is not set yet or
-            // - the timeout task cannot schedule yet or
-            return;
+    public boolean scheduleTimeout(long timeoutMillis) {
+        checkArgument(timeoutMillis > 0,
+                      "timeoutMillis: " + timeoutMillis + " (expected: > 0)");
+        // Do nothing if the timeout was scheduled already
+        if (timeoutFuture != null) {
+            return false;
+        }
+
+        cancelTimeout();
+        if (!timeoutTask.canSchedule()) {
+            return false;
         }
 
         this.timeoutMillis = timeoutMillis;
-        firstStartTimeNanos = lastStartTimeNanos = System.nanoTime();
-        if (timeoutMillis > 0) {
-            timeoutFuture = eventLoop.schedule(timeoutTask, timeoutMillis, TimeUnit.MILLISECONDS);
+        final long nanoTime = System.nanoTime();
+        if (firstExecutionTimeNanos == 0) {
+            firstExecutionTimeNanos = nanoTime;
         }
+        lastExecutionTimeNanos = nanoTime;
+        timeoutFuture = eventLoop.schedule(timeoutTask, timeoutMillis, TimeUnit.MILLISECONDS);
+        return true;
     }
 
+    /**
+     * {@inheritDoc}
+     *
+     * <p>Note that the {@link TimeoutTask} should be set via the {@link #setTimeoutTask(TimeoutTask)} or
```

Comment:
It seems like there's not much chance that a new `TimeoutTask` it set.
How about removing the setter and setting it in the constructor?
```
// In HttpResponseDecoder
HttpResponseWrapper(...) {
    super(eventLoop, newTimeoutTask(delegate, ctx));
    this.delegate = delegate;
    this.ctx = ctx;
    this.maxContentLength = maxContentLength;
    this.responseTimeoutMillis = responseTimeoutMillis;
}

// In HttpResponseSubscriber
HttpResponseSubscriber(ChannelHandlerContext ctx, HttpObjectEncoder responseEncoder,
                       DefaultServiceRequestContext reqCtx, DecodedHttpRequest req,
                       boolean enableServerHeader, boolean enableDateHeader) {
    super(ctx.channel().eventLoop(), newTimeoutTask(reqCtx));
    this.ctx = ctx;
    this.responseEncoder = responseEncoder;
    this.req = req;
    this.reqCtx = reqCtx;
    this.enableServerHeader = enableServerHeader;
    this.enableDateHeader = enableDateHeader;
}
```

Reply:
That is a nice suggestion. I tried to make `TimeoutTask` immutable in `DefaultTimeoutController` initially. We have to make `newTimeoutTask` static to call `newTimeoutTask` in `super`. But it is difficult because it refers to the `state` that is non-static field.

Or there is something I missed, let me know :-)